### ISSUE TITLE
Time stamps are all wrong with variable frame rate video

### ISSUE THREAD

**Issue by LukeM-Github** [Breakthrough/PySceneDetect#168]
I glanced at the code, it seems to be assuming constant frame rate.

**Comment by Breakthrough** [comment]
This is correct.  Adding support for variable frame rate videos will really drive a lot of underlying API changes, which I'm not opposed to, but will require to rethink how timestamps are stored within PySceneDetect internally to achieve this in an efficient matter.

Any suggestions on how this can be achieved are most welcome.

**Comment by Breakthrough** [comment]
I think I have a good initial design, will post here when there are some unit tests showing how timecodes will be stored in v1.0.  The idea is to store both a frame number, time in seconds, framerate, and a tag if the framerate is variable or not.  Unit tests should make this much more clear, so will post an update here when they are ready in the 1.0 branch.

**Comment by Breakthrough** [comment]
Timecodes as of v0.6 are now obtained directly from the underlying VideoStream (via OpenCV VideoCapture or PyAV), but timecode calculations still assume a constant frame rate.  This should improve accuracy for the majority of use cases, but there still remains a significant amount of work to be done for supporting VFR material.

Anyone willing to share samples for testing is encouraged to upload them here to this issue for analysis.

**Comment by hokaso** [comment]
> Timecodes as of v0.6 are now obtained directly from the underlying VideoStream (via OpenCV VideoCapture or PyAV), but timecode calculations still assume a constant frame rate. This should improve accuracy for the majority of use cases, but there still remains a significant amount of work to be done for supporting VFR material.
> 
> Anyone willing to share samples for testing is encouraged to upload them here to this issue for analysis.

I have a better idea, that is, first transcode the vfr video into a smaller size cfr video, then perform scene detection on the cfr video, and then apply the cfr timeline information to the vfr video. I have tested, is completely no problem.

**Comment by Breakthrough** [comment]
> I have a better idea, that is, first transcode the vfr video into a smaller size cfr video, then perform scene detection on the cfr video, and then apply the cfr timeline information to the vfr video. I have tested, is completely no problem.

I fear this would greatly impact performance, and it should be able to get timing information from the underlying video I/O library.

**Comment by Breakthrough** [comment]
I think a general overview of how to land this:

1. Figure out how to get correct timing into VideoStream.
    - We need to store the frame number, PTS, and timebase
3. Combine timing info with frames rather than querying separately in VideoStream.
    - We don't need the "advance" arg in the read method
    - Combine timestamps with frames so they are returned together
4. Integrate with Detectors + SceneManager.
5. Handle user input for start/end time, min scene length, etc.

**Comment by Breakthrough** [comment]
Fixed in main, will try to get a beta release of 0.7 out as soon as possible.

**Comment by Fmstrat** [comment]
@Breakthrough I noticed 0.7 was released, but dvr-scan hasn't had a release (on GitHub) to include this, but the `pip` version seems to have been updated to .1 which might?

I'm using `1.8.2.1` from `pip`, but my variable bitrate AVI (20fps in the header, but 40fps average from `ffprobe`) plays any extracted motion clips back as 40fps (about twice as fast as it should).
